### ISSUE TITLE
Species pages and mapping functionality

### ISSUE THREAD

**Issue by Sam-Amy** [BiologicalRecordsCentre/iRecord#1286]
Following our discussion (20.04.22) about iRecord data visualisation tools produced by Rich Burkmar, re: linking the new species mapping functionality with the existing species page containing info/photo. 

Issues with current species page layout (e.g. https://irecord.org.uk/species-details?taxon_meaning_id=71446): 
- Very hard to find (must go via species record at present); can this be made easier to navigate to using a url containing some more easily referenceable identifier such as TVK? 
- Doesn't show full info on left hand side (cropped by photos). 

- Photos are displayed vertically; is a gallery view possible? 

- Some have huge lists of common names; could these displayed only on request via a button/pop-up?  

In terms off implementing the mapping functionality, would a BSBI-type format work, with a page for species info with static inset Atlas-style image, and a link above this to interactive mapping (e.g. https://www.brc.ac.uk/plantatlas/plant/orchis-mascula)?  

Raised here as an issue for discussion in next iRecord prioritisation meeting.

**Comment by johnvanbreda** [comment]
A couple of extra notes for discussion:
* You can also access the species page via the Species tab on the Explore pages. But, I agree if the page was improved, then it makes sense to make it more widely accessible from other places where species names are shown.
* The page's styling didn't get done when it was copied from Drupal 7 I think - this would be an easy fix.
* The page is often very slow to load but I think that is because it tries to load a full set of photos. Photos should be paginated and it would be nice if some prioritisation based on quality of the image could be applied.

Making it work from TVK is definitely possible.

**Comment by burkmarr** [comment]
> The page's styling didn't get done when it was copied from Drupal 7 I think - this would be an easy fix.

@johnvanbreda - where would the CSS for the page have been stored?

**Comment by burkmarr** [comment]
I guess it could have been in a file referred to in the **Additional CSS files to include** section on the form. @BirenRathod - is there still a version of the Drupal 7 site available where I can check this out?

**Comment by BirenRathod** [comment]
@burkmarr I send yop email about it.

**Comment by BirenRathod** [comment]
@burkmarr it stores in node.xx.css. On drupal 9 -> sites/default/files/indicia/css 
On drupal 7 -> client_helpers\prebuilt_forms\css

**Comment by BirenRathod** [comment]
> > The page's styling didn't get done when it was copied from Drupal 7 I think - this would be an easy fix.
> 
>  - where would the CSS for the page have been stored?
@burkmarr  which file or page are you after?

**Comment by burkmarr** [comment]
@johnvanbreda - thanks for the tip about comparing to `record_details_2.php`. That also helped with figuring out that the tabulation problems can be fixed by replacing this line:
```php
'bands' => array(array('content'=>$indicia_templates['dataValue'])),
```
with this:
```php
'bands' => array(array('content' => str_replace(['{class}'], '', $indicia_templates['dataValue']))),
```
Your notes on the image markup sorted out their basic formatting. With those two fixes the page looks like the image below on my dev setup (for species-details?taxon_meaning_id=89659).

Neither these images or those displayed by the  `record_details_2.php` have the nice formatting previously used on the Drupal 7 site (see second image below for comparison). In the D7 theme there was CSS for the `.gallery-item` class that provided this formatting (`/sites/all/themes/irecord/css/irecord.css`) which doesn't appear to have been moved into the D9 project. @BirenRathod - is that something that you can sort out?

![image](https://user-images.githubusercontent.com/9275178/173105417-800e2c11-6226-4c54-80e3-97266a966799.png)

Images from the old D7 site:

![image](https://user-images.githubusercontent.com/9275178/173106160-2776ff0b-25ed-4419-bb8f-ceca38986daa.png)

**Comment by burkmarr** [comment]
Forgot to mention a problem which I can't resolve with the page title. The title for the taxon above looks like this:

![image](https://user-images.githubusercontent.com/9275178/173107841-1b5d4706-3f74-4f5d-8756-bbc15e74eff0.png)

It looks like Drupal 9 doesn't interpret HTML tags in the title. I looked at the `iform_preprocess_page_title` hook in `iform.module` and researched the problem on Google, but couldn't find a way to fix it. The only option seems to be to strip out the `<em>` tags in `species_details.php` so it looks like this:

![image](https://user-images.githubusercontent.com/9275178/173108673-9ca018c9-ca5d-437e-898c-1a233ebf35f4.png)

Not ideal, but I can't see an alternative.

**Comment by kitenetter** [comment]
This is a side-issue, but I'm not convinced that it is a good idea to display all the Pantheon data unless we can provide quite a lot more context and links to the Pantheon help pages. The ones that do make sense without further context are:
- adult guild
- larval guild
- broad biotope
- habitat

Similar questions arise for the plant taxon attributes.

**Comment by BirenRathod** [comment]
@burkmarr Here is css. 

.gallery-item, .gallery li, .summary-report {
  text-align: center;
  padding: 0.5em;
  margin: 0.5em;
  display: inline-block;
  vertical-align: middle;
  *vertical-align: auto;
  *zoom: 1;
  *display: inline;
  list-style-type: none;
  background: none;
  -moz-box-shadow: silver 3px 3px 6px;
  -webkit-box-shadow: silver 3px 3px 6px;
  box-shadow: silver 3px 3px 6px;
  border: solid silver 1px;
}

.gallery li button {
  overflow: hidden;
}

**Comment by burkmarr** [comment]
@BirenRathod - yes I saw that. But where's the best place for it? I'm not sure about the theme architecture of the new site.

**Comment by burkmarr** [comment]
@kitenetter 

> I'm not convinced that it is a good idea to display all the Pantheon data...

Mabey that can be fixed via the report it runs - I'll take a look tomorrow.

**Comment by BirenRathod** [comment]
>  yes I saw that. But where's the best place for it? I'm not sure about the theme architecture of the new site.
@burkmarr you can add it to themes\custom\irecord\css\style.css.
If it is inidicia node specific then you can create node.x.css (x is number of that node) and upload via sftp into files/indicia/css folder.

**Comment by burkmarr** [comment]
@kitenetter 

> I'm not convinced that it is a good idea to display all the Pantheon data...

> Mabey that can be fixed via the report it runs - I'll take a look tomorrow.

My mistake - fields can be excluded at the form configuration level, so that was easier to take care of than I thought.

**Comment by burkmarr** [comment]
@johnvanbreda

> The page is often very slow to load but I think that is because it tries to load a full set of photos. Photos should be paginated and it would be nice if some prioritisation based on quality of the image could be applied.

It looks like the slowness was due to the `@noGeoserver` option for the map control being set to `true` causing the page to download a vector representation of all the records which is too slow for common taxa. I've changed the value to `false` to use GeoServer - which prevents the hang but there is a delay before the map is displayed. I'm not sure why that is - there's very little delay on the species-maps pages which also use GeoServer - so I'll try to track the difference down.

There is already a `@itemsPerPage` option on the photos control whose default is 20, but no pagination control is displayed. Is there another page that has paginated images that I can look at?

**Comment by johnvanbreda** [comment]
@burkmarr looking at the formatting of the Species Details section, I can see that there is a replacement token in the HTML template `{class}` which is not being replaced. This causes the browser to not interpret the `<dd>` elements correctly, breaking the layout. The equivalent code for the 'bands' parameter to the freeform report in record_details_2.php looks like this, which fixes the issue:
```php
  'bands' => [['content' => str_replace(['{class}'], '', $indicia_templates['dataValue'])]],
```

Looking through the list of attributes, many of the Pantheon ones might be of interest (larval, adult guild etc) and there are lots of attributes from other systems as well, so maybe we need to create a list of generally useful attributes and only show those.

There is no pagination control on photos because the code also sets limit to 20. I think the intention here was to prevent a full count on the server-side which might be slow. Is there really a need to paginate back through all the images? In the absence of a quality based prioritisation or similar, maybe just showing a bunch of recent example images is enough?

**Comment by burkmarr** [comment]
@johnvanbreda - I had already found the `{class}` problem (see https://github.com/BiologicalRecordsCentre/iRecord/issues/1286#issuecomment-1152521355).

> Looking through the list of attributes, many of the Pantheon ones might be of interest (larval, adult guild etc) and there are lots of attributes from other systems as well, so maybe we need to create a list of generally useful attributes and only show those.

@kitenetter - would you be able to come up with such a list?

> Is there really a need to paginate back through all the images?

Not from me John - I just made the assumption that it was a feature because I misinterpreted the name of the option `itemsPerPage`. Suggest we that as it is for now. I'd like to get the page working as initially intended in order that @kitenetter and @Sam-Amy can better evaluate it before we decide how to enhance it.

**Comment by burkmarr** [comment]
> there is a delay before the map is displayed. I'm not sure why that is - there's very little delay on the species-maps pages which also use GeoServer - so I'll try to track the difference down.

This turned out to be because the distribution layer feature type was set to `indicia:cache_occurrences` - changing it to `indicia:cache_occurrences_functional` speeds it up.

**Comment by burkmarr** [comment]
@johnvanbreda - there are a few  small changes to fix the immediate formatting and performance issues with the page that affect these two files:

- web/modules/custom/iform/client_helpers/prebuilt_forms/species_details.php
- web/themes/custom/irecord/css/style.css

I'm uncertain of the workflows now that we've moved to Pantheon. For the theme update, shall I just push that on the Pantheon repo? And for the client_helpers update, should I just do what I used to do with the Drupal 7 site - i.e. branch from dev and push to GitHub for you to review and merge as you see fit?

**Comment by BirenRathod** [comment]
@burkmarr for style.css, yes just push the code on Pantheon repo.

**Comment by johnvanbreda** [comment]
@burkmarr are the changes to the theme style.css actually iRecord theme specific, or would the Indicia code generally benefit from them?

For the other changes, as these are actually fixes rather than developments, they should be hotfixed (i.e. branch off master, then merge into master and develop). If you would like me to review them then you can do a pull request on the master branch but if you are comfortable that the changes are non-controversial you can just merge directly. 

Once the hotfix is merged to master, the affected files can be updated on the brc-irecord Pantheon repository.

**Comment by kitenetter** [comment]
@burkmarr are we aiming for one list of species attributes for **all** species, or can we have different attributes for different taxon groups (as we currently do, e.g. in your examples of leafhopper vs plant).

And I'm not sure what the full list of potential attributes would be. Candidates appear to be:
1. The conservation statuses listed in taxon_designations (do you also need this linked to taxa_taxon_designations?)
2. taxa_taxon_list_attributes

Can we use both of those sets of attributes? What do you need, a list of IDs from one of these tables?

**Comment by burkmarr** [comment]
> are the changes to the theme style.css actually iRecord theme specific, or would the Indicia code generally benefit from them?

@johnvanbreda they provide a border for images where class `media-gallery` is specified as happens in a number of client_helper forms (5 that I can see). 

This:

![image](https://user-images.githubusercontent.com/9275178/173390101-59f59eaf-9d49-4e12-bafe-2b08b22e336b.png)

to this:

![image](https://user-images.githubusercontent.com/9275178/173390249-d1aa6ed4-5af2-4f76-9b41-aa2a0b408285.png)

**Comment by kitenetter** [comment]
Okay, so if we want one list for all taxa I'd suggest
- synonyms
- adult guild
- larval guild
- broad biotope
- habitat

And I was going to add some suggestions from the plant attributes as well but it's difficult to know what to select without seeing how they appear on the page. One answer might be to provide a link to the species page on the plant atlas, where the Ellenberg info is displayed in a more helpful context?

The [plant example](https://irecord.org.uk/species-details?taxon_meaning_id=71446) has what seems to me to be an unhelpfully long list of vernacular names. I'd prefer to see one English name, and perhaps a toggle for "show all other vernacular names", but not sure if the species dictionary enables a 'best' English name to be separated out.

Sorry, not sure I'm helping much here, but the more I think about this the less I like the idea of throwing out lots of attributes that happen to be in the database without providing enough context to make them understandable. I think we either need to spend time developing a page that will provide enough context, or leave them out.

Happy to have a zoom chat if it helps. Or if I'm just over-complicating things, and the priority is to get a page working, then I guess we can just display all the attributes under a "more info" toggle and leave it there until such time as we can improve the display.

**Comment by burkmarr** [comment]
@kitenetter - yes there's a lot to consider so maybe that should come within the context of the further development or redevelopment of the page. So let's get the page - as it currently stands - looking right and then think about where it needs to go. I need to do some thinking/investigation about the best way forward technically with integrating the new visualisations.

**Comment by burkmarr** [comment]
@johnvanbreda - I applied the hotfix as you described. Is the workflow now to update the iRecord pantheon site with the new client_helpers master branch? If so, shall I do that?

**Comment by BirenRathod** [comment]
@johnvanbreda Can you also include the amendments you done for species details page with this hotfix?

**Comment by johnvanbreda** [comment]
@BirenRathod which amendments? I think the species details page on iRecord should be up to date apart from applying Rich's hotfix.

@burkmarr yes, you can do that, but if applying all the files I like to double check all the changes before committing and pushing them. In theory there won't be any other changes, but line ending issues have made this very tricky to check in the past, though hopefully these have been sorted now. As this hotfix is isolated, you might prefer to just manually patch the changed lines of code which theoretically has the same end result.

**Comment by burkmarr** [comment]
@kitenetter and @Sam-Amy the species details page is now displaying much more as originally designed. I'm going to investigate mapping options etc and I suggest we pick this up again once I've done that.

**Comment by Sam-Amy** [comment]
Thanks Rich – it looks much better already! I am awe of all your skills @burkmarr<https://github.com/burkmarr> and @johnvanbreda<https://github.com/johnvanbreda>, I must admit most of what you have been discussing has gone way over my head, but fascinating none-the-less!

I’m sure this is part of everything you are already looking at, but in case it is more of a problem on the tiny laptop I’m working on today…many of the species detail headings are cut off (e.g. ‘Ellenberg indicator v…’) rendering the associated numbers meaningless (screenshot below). I guess a separate page to click/open up from the main species/map page would give more room and rectify this?

***@***.***

**Comment by johnvanbreda** [comment]
@BirenRathod you can commit the changes for the redirect control to master and develop using a hotfix and then to the iRecord repository if you are waiting for it. That's separate to the species details changes though.

**Comment by burkmarr** [comment]
@Sam-Amy 

> I’m sure this is part of everything you are already looking at, but in case it is more of a problem on the tiny laptop I’m working on today…many of the species detail headings are cut off (e.g. ‘Ellenberg indicator v…’) rendering the associated numbers meaningless (screenshot below).

Noted - if we end up displaying these attributes, we'll  have to address this in some way. (It's a problem on my desktop screen too.)

**Comment by BirenRathod** [comment]
@Sam-Amy & @burkmarr Can you both share some screenshot here please? I don't see any screenshot here. I need to figure out the problem od general styles of iRecord and indicia.

**Comment by BirenRathod** [comment]
Thanks, @Sam-Amy.

**Comment by burkmarr** [comment]
@johnvanbreda - I want to either change a report (https://github.com/Indicia-Team/warehouse/blob/master/reports/library/taxa/taxon_names.xml) create a new version, e.g. taxon_names_2.xml, that allows a search on external key (tvk) as a further alternative to searching on taxon_meaning_id or taxa_taxon_list_id and also return the default_common_name field. I have created it (library/taxa/taxon_names_2.xml) on the dev warehouse and it's working fine. I have two questions:

1. Which approach is best - update the existing report or create the new version? I don't think that the changes would break anything which already uses the report.
2. Whatever your answer to 1, what is a good workflow for me to deploy it to the live warehouse?

**Comment by burkmarr** [comment]
@kitenetter @Sam-Amy - the species details form was not eliminating duplicates from either the common names or synonym arrays. Once I've taken care of that and also identified the `default_common_name`, the top section for Early-purple orchid looks like this:

![image](https://user-images.githubusercontent.com/9275178/180232751-88be2376-ca05-49a0-9cc2-a3aba4a7be67.png)

instead of this:

![image](https://user-images.githubusercontent.com/9275178/180233083-ad9b3248-7bfb-41ed-919f-13819e356d35.png)

That looks okay to me without doing anything flashy like hiding other common names unless selected. Do you agree?

I'm going to delve a bit more into the custom attributes to see what possible approaches there might be to deal with them.

**Comment by Sam-Amy** [comment]
Looks good to me too.

**Comment by kitenetter** [comment]
Agree, that's much easier to interpret.

**Comment by johnvanbreda** [comment]
@burkmarr I think the correct approach is a hotfeature branch off master, then merge into master and develop. Let me know if you need more details.

**Comment by burkmarr** [comment]
@kitenetter  and @Sam-Amy - I've done some work on this based on previous comments in the thread. These are implemented in a new prebuilt form (species_details_2) which I've deployed on a Pantheon branch and used in a new species details page (species-details2) e.g:

https://burkmar-brc-irecord.pantheonsite.io/species-details2?taxon_meaning_id=71446

It can be compared to the current version which is still available (species-details) e.g:

https://burkmar-brc-irecord.pantheonsite.io/species-details?taxon_meaning_id=71446

We can use this branch to test ideas and guide the development. The branch is using the dev warehouse in order to pick up a new report required for the new page. Unfortunately many of the images are not available there - so don't worry about the image thumbnails not appearing as expected.

Some notes to help you evaluate:

- Duplicate names removed from common names and synonyms.
- The 'default' common name is listed as 'common name' separately from 'other common names'.
- In the example above, I've utilized the existing form feature to list only specified custom attributes (see below for more info).
- If custom attributes with long names are included, the names are word-wrapped instead of being truncated (see below).
- I've added an `external_key` (TVK) URL parameter to the existing `taxon_meaning_id` and `taxa_taxon_list_id` parameters so that the page can be reached by specifying a TVK in the URL.
- I've added three new mapping controls - `[hectadmap]`, `[exploremap]` and [`dualmap`]. The `[exploremap]` control just uses the existing leaflet grid map deployed on other pages such as the explore all records page. The `[dualmap]` control displays both the [`hectadmap]` and `[exploremap]` control with a switch to toggle between them.
- I've added a new '[recsbyyear]` control to show the temporal distribution of records.

I couldn't think of a way to improve on the existing methods of limiting the custom attributes displayed. If you want to experiment with this, edit the form and alter the attributes in the form configuration section marked FIELDS FOR SPECIES DETAILS. In the form as currently configured you need to add the attribute caption to the list there in order to display it for any taxon. If the attribute is not present for any particular taxon, the caption is not shown.

**Comment by johnvanbreda** [comment]
@burkmarr looks good to me. A couple of thoughts:
1. Remove duplicate attribute labels when there are multiple values of the same attribute (BR Habitats in this case).
2. As well as the temporal distribution chart by year, a phenology of month in year would be good?
3. As the map is static, consider the proxyCacheTimeout option (https://indicia-docs.readthedocs.io/en/latest/site-building/iform/helpers/elasticsearch-report-helper.html#advanced-options).
4. Do we need the Explore map?

**Comment by DavidRoy** [comment]
Thanks @burkmarr 
2. I agree that a phenology chart would be a good addition
4. also agree with this as the page has an explore records link

The species name looks odd with a double authority?
Orchis mascula (L.) L.

**Comment by Sam-Amy** [comment]
Looks great @burkmarr. Apologies if I'm missing something obvious or already mentioned, but why are there so few records showing on the 'Explore records' map, compared with the map on the original page here: https://burkmar-brc-irecord.pantheonsite.io/species-details?taxon_meaning_id=71446 ?

**Comment by burkmarr** [comment]
@Sam-Amy - it's a different style of map. The old one displays the actual record locations from a geoserver WMS - the new one (which is the same as the grid map on the explore records page) displays records aggregated to atlas squares (e.g. hectads, tetrads, monads) at lower zoom levels but at high zoom switches to the record points. In any case it's a moot point because the concensus seems to be not to include the interactive map because there is a link to take you to the full explore records page anyway.

**Comment by Sam-Amy** [comment]
Thanks @burkmarr 

When I was looking earlier I could see just a few points dotted around the UK, which didn’t make sense according to this. I can now see far more, so I’m not sure what was going on there?!

**Comment by burkmarr** [comment]
To address the points above, I have:

1. Removed duplicate custom attribute names.
2. Added a phenology chart.
3. Made use of the proxyCacheTimeout option on ES queries.
4. Replaced the dual explore map with the hectad map only.

https://burkmar-brc-irecord.pantheonsite.io/species-details2?taxon_meaning_id=71446

The phenology chart (`[recsthroughyear]`) has a '@period' option that can be set either to 'week' or 'month' - this doesn't change the x axis (which always shows months) but it affects the way the line is generated (52 points of weekly totals vs 12 months of monthly totals) and changes the y axis and label accordingly.

@johnvanbreda - I set the proxyCacheTimeout option to 300 for the ES queries. Is that reasonable?

The dual map is still available on the form (control `[dualmap]`) but I have replaced it with the `[hectad]` control in the form configuration.

@DavidRoy - I also thought that the authority on 'Orchis mascula (L.) L.' looked strange and thought that it was being mangled somehow. But when I looked at the database, that is how the authority is specified. I think it might signify that Linnaeus originally named the taxon differently and then renamed it *himself* to Orchis mascula.

**Comment by johnvanbreda** [comment]
> I set the proxyCacheTimeout option to 300 for the ES queries. Is that reasonable?

Yes, that seems fair to me.

**Comment by burkmarr** [comment]
@kitenetter I will do something to separate the map from the stuff below it.

I think that adding a facility to change the displayed hectads based on date of the records would be useful in the context of the species summary page, so happy to look at this. There are two obvious ways to do it:

1. Apply a user-controlled year range filter to the records that are used to construct the hectad atlas map. That would result in a map that only shows hectads that have at least one record in the specified period. Since colour would not be required to indicate a date band, it could still be used to give an indication of the relative number of records in each hectad.

3. As you've described, apply user-controlled year threshold(s) to change the colour of any particular hectad, indicating which year band it is in. (Presumably the year band assigned to a hectad would be determined by the most recent record in it.)

I'm happy to look into wichever of these you think is the best approach. If the second approach, then we need to decide on the number of year bands. Two year bands would only require a single control for the user to specify the threshold year. If more than two bands then more controls need to be added - complicating the user interface.

I agree that the life-stage feature would be useful, but because of the amount of flexibility in way this field is used across taxonomic groups (and even within groups), it is not straight-forward to apply on a general species details form. The list below shows the top 50 terms used for the iRecord site for data over the last 5 years. (The count for the term 'not recorded' is for those records where this is explicitly entered. Most records have `null` recorded against this field which isn't included in the table.)

So even picking out something as simple as adult counts is problematic. From just the terms listed below, you would probably want to include 'adult', 'Adult', 'Adults' and 'adult bug' as well as deciding whether or not 'not recorded' and null values should be assumed to be adults. I guess the way to handle this would be in the page configuration - much like the custom attributes discussed above (the problems are similar in many ways). But we have seen that this isn't ideal. Then there is the problem of what options to offer the user on the page itself - options relevant for invertebrates would not be relevant to many other taxonomic groups for example. 

occurrence.life_stage.keyword:   Descending | Count
-- | --
adult | 1,795,383
Adult | 844,166
not recorded | 617,843
pre-adult | 47,842
flowering | 37,048
other | 16,405
vegetative | 9,603
larva | 8,393
native | 6,083
nymph seen in spittle foam | 5,353
gall | 4,228
nymph | 4,162
gametophyte (not fruiting) | 4,044
Nymph | 3,339
Vegetative | 3,196
Larva | 2,970
fruiting | 2,933
spittle foam only | 2,418
mine (empty) | 2,405
Not recorded | 2,284
mine (tenanted) | 2,024
mine | 1,973
sporophyte present (fruiting) | 1,781
immature | 1,752
Uncertain | 1,587
Adults | 1,308
Egg | 879
Plant bare | 683
pupa | 640
egg | 534
Immature | 524
In leaf | 523
young | 500
dead | 468
Flowering | 464
introduced - accidental - regenerating | 451
In flower | 446
Pupa | 417
adult bug | 404
sporophyte | 386
Other | 380
Young | 375
introduced - intentional | 372
Juvenile | 370
Native | 345
larval web | 340
Leaf-mine | 333
larval case | 288
With ripe fruit | 259
spawn/egg | 204

**Comment by kitenetter** [comment]
I agree that the life-stage issue is complex, and taxon-group specific, so let's leave that aside for now.

For date banding my preference would be your option 2, with user-defined date bands. Ideally I'd go for three time periods, with defaults set to:
- before 1980
- 1980 to 1999
- 2000 onwards

But if three time periods is going to add too much complexity then having two time periods would still be good, and I'd suggest defaulting to before 2000, and 2000 onwards (based on most recent record in the hectad - although enabling users to switch to oldest record in the tetrad would also be interesting, e.g. for mapping spread of invasive species ...).

If it's possible to allow users to change the defaults that would be great. 

As always I'm conscious that this is just my opinion, and other users may want different things.

**Comment by DavidRoy** [comment]
As ever, your opinion is good @kitenetter.  I would support these three time periods options as a start.  A more flexible date-range picker can be a future requirement?

**Comment by burkmarr** [comment]
Okay great. I will proceed with three coloured time periods.

**Comment by burkmarr** [comment]
I have implemented the coloured time bands. The three colours can be changed via the page configuration if required. (For the defaults I've used three - 'colourblind safe' from https://colorbrewer2.org/.)

The user sets the thresholds for the time bands using a couple of numeric text controls. I conisidered trying to do something more flashy with sliders, but that would take more effort and is perhaps one for the future? In any case I think this works pretty intuitively - the user can immediately see the effects of the settings they use by referring to the map legend which is displayed immediately below the controls. I've put some logic in there to prevent nonesense values being set (e.g. threshold 1 greater than threshold 2).

It was also easy for me to add the select control which allows the user to toggle between 'most recent on top' (the default) and 'oldest on top'. I thought this could be useful, but if you think the complication to the UI is too much I can remove it (or make its inclusion configurable).

Also it occurred to me that a useful feature would be mouse-over info for the map dots (the map control already supports it and it is a useful feature on the BSBI Atlas site). So if you look below the map you can see how you can get the hectad name, number of records, min record year and max record year for any dot when the mouse is over it.

https://burkmar-brc-irecord.pantheonsite.io/species-details2?taxon_meaning_id=71446

**Comment by burkmarr** [comment]
I forgot to say - the initial values for the thresholds are configurable, so easy to change.

**Comment by DavidRoy** [comment]
looks good to me.
One small thing.  Can the label 'BR Habitats' be changed to 'Broad Habitats'

**Comment by kitenetter** [comment]
@burkmarr that looks brilliant and functions brilliantly! For the default year thresholds I'd suggest 1999 and 2009:

![image](https://user-images.githubusercontent.com/16039111/185204450-043a9602-9e90-4b2a-b3e0-2ad729316b52.png)

Can we re-word the key:
- "Most recent record in hectad" becomes "Year of latest record in 10 km square"
- "Oldest record in hectad" becomes "Year of first record in 10 km square"

I think you are including unverified records, and excluding rejected records - it would be good to say so somewhere on the page.

One more mission-creep feature request: is it possible to enable the three date bands to be switched on or off individually, so that I could generate three separate maps for three separate time periods, e.g. to use in a presentation?

Just for fun, here's an example of an expanding species (Downland Villa bee-fly), using the "oldest" display:

![image](https://user-images.githubusercontent.com/16039111/185205096-92a45aa0-1d60-4018-9b4f-aec023283458.png)

**Comment by burkmarr** [comment]
@kitenetter - great, I'm glad that it fits the bill. I've added some text re the rejected and unverified records under the map (let me know if you'd  prefer a differet set of words) and changed the legend titles as requested. I've had a think about enabling/disabling bands and I saw the similarities with the highlighting feature on other charts (e.g. donut chart) and I thought it could be an useful general feature to add to the map library. It should be quite easy to do I think so I will have a look at it today and/or tomorrow.

**Comment by burkmarr** [comment]
@DavidRoy 

> Can the label 'BR Habitats' be changed to 'Broad Habitats'

The label is the caption for that taxon attribute as defined in the Warehouse - https://warehouse1.indicia.org.uk/index.php/taxa_taxon_list_attribute/edit/76. It could either be changed in the Warehouse, but that would affect it wherever it is used. Alternatively I could add configuration to the species_details_2 prebuild form which allows custom attribute names to be 'translated' for the purposes of this form only.

**Comment by DavidRoy** [comment]
@burkmarr I think over-riding the warehouse attribute name is the best option.  I could imagine we might want to do this for other information we display in these sorts of pages

**Comment by burkmarr** [comment]
@DavidRoy - just to clarrify, do you mean leave as is in the Warehouse but add some configuration capability in the page to overrride it?

**Comment by DavidRoy** [comment]
configuration to the page

**Comment by burkmarr** [comment]
> is it possible to enable the three date bands to be switched on or off individually, so that I could generate three separate maps for three separate time periods, e.g. to use in a presentation?

@kitenetter I have implemented a feature whereby a user can click on the legend on the map to highlight/lowlight the sets of dots that each element on the legend relates to - effectively meeting the requirement above. The highlight and lowlight styles are configurable. That means, for example, that if you would rather that the non-relevant dots are completely hidden (rather than faded as at present), this can be achieved through the `@lowstyle` control config parameter for the page. Likewise the highlight options currently change the colour to black but this could be changed through the `@highstyle` control config parameter (e.g. set the parameter to nothing to leave the dots with their original colour). 

https://burkmar-brc-irecord.pantheonsite.io/species-details2?taxon_meaning_id=71446

A user will probably not find this feature by accident, so it could probably use a brief explanation somewhere on the page (e.g. 'Click on legend to highlight ranges.').

The feature will be useful in other contexts where the atlas map is used (e.g. potentially the BSBI atlas) - so thanks for the idea!

@DavidRoy after speding some time looking into the code, I eventually resorted to a workaround to achieve the custom attribute label override. The configuration is easy enough to handle, but actioning properly would require modification of reporting code deep in Indicia which would require more investigation and learning on my part (and testing since modifying it would potentially affect many parts of Indicia websites). So for now I've done the modification using Javascript after the Indicia code has generted the table. (Actually that's also the method I used to do the duplicate label hiding.) It works well enough, but it's not technically the best solution for these two requirements.

**Comment by DavidRoy** [comment]
Thanks @burkmarr. I'd be very happy for this to now go live.  We can then move onto other visualisation pages?
@kitenetter what do you think?

**Comment by kitenetter** [comment]
Agree it should go live, it's a great enhancement for the site.

I think the configuration for the single element in the key is okay - we could debate different ways of doing this but what we have is a perfectly good starting point. As you say, it will need some text to explain how to use it. Suggest:
- Click on one of the date ranges in the key to highlight that particular range
- Click anywhere else on the map to go back to all three date ranges

One more suggestion (but this shouldn't delay going live): I've already started using these maps for soldierflies on social media, and to do so I'm just screen-shotting the map area. Maybe we could repeat the taxon name above the key, and add an iRecord logo under the key, just so that if the map is picked up as a stand-alone graphic it gives it some context?

**Comment by burkmarr** [comment]
@kitenetter - I've added an info text block that appears above the map whose text can be set with a `@info` page config parameter for `[hectadmap]` and I've set that to text you suggested above on the demo page.

Dealing with variable text in a map legend is problematic because of layout and space considerations. Some taxa have very long names and would not be easily accommodated. Instead I've taken advantage of an existing feature of the atlas map library which is a download feature that enables some text to be added to the foot of the downloaded image. I've added a `@download` page config param for `[hectadmap]` which, if set to true, causes a download button to be displayed, enabling the map image to be downloaded without doing a screen-shot. Text is added to the foot of the image which specifies the taxon name as well as where and when the image came from (and repeating the text about rejected records excluded and unverified included).

![image](https://user-images.githubusercontent.com/9275178/188857581-d9be8f86-e198-4e77-9b97-da446a4d9774.png)

I know that the taxon name is not prominent as it would be if it was in the legend title, but it does at least ensure that the provenance of the image is always attached to it.

Adding an iRecord image logo is also problematic for several reasons and would require some work. Adding one underneath the text added to the foot of the downloaded page is already possible with the atlas map library but it would have to be made configurable because, in theory, this form could be used on other Indicia sites, not just iRecord. If it's okay with you I think that we should leave this for now so that I can press on with getting the code reviewed by @johnvanbreda in some pull requests so that we can proceed to publishing on the live site.

**Comment by burkmarr** [comment]
@kitenetter - I should also mention another limitation of the image download - it doesn't reflect the interactive highlighting. So if you wanted an image showing highlighted dots, you'd still have to screen grab.

**Comment by kitenetter** [comment]
Thanks @burkmarr - the text block that you've added provides clear metadata so I'm happy with that.

**Comment by burkmarr** [comment]
@johnvanbreda - what is the best way to filter ES to all records of one taxon? I had thought that `taxon.taxon_meaning_id` would do it, but when I looked at records for *Polygonia c-album* I saw several values for this field. So I looked at `taxon.accepted_taxon_id` and saw that even that could have more than one value (NHMSYS0000503893 and eBMS446).

I settled on filtering on `taxon.accepted_name`. I think that this should normally be fine, but I found a problem with *Aglais io* - when this is indicated in the URL for the new taxon page (https://irecord.org.uk/species-details?taxon_meaning_id=127683) the XML report returning all the taxon names from `cache_taxa_taxon_lists` indicates that *Inachis io* is the preferred taxon name rather than *Aglais io* so when I then filter ES on that name, I get no records.

**Comment by johnvanbreda** [comment]
The UK based data should all map to UKSI via the search_code (the name TVK) and the external_key (the accepted name TVK) so that's normally the best way (or even the recently added Organism Key which is more stable if the accepted name changes). The problem here is that you are trying to view all records from both the UK and the European sections of the system and the keys don't match - they have to this point really been 2 completely separate parts of the system. That's why you are finding 2 different accepted_taxon_id values.

Taxon meaning ID provides a theoretical way to join taxa across lists/taxonomies, but it is not likely to give the correct result in practive.

Accepted name will only work if the accepted name doesn't change, as you've discovered. I remember Charles Hussey (Chris Raper's predecessor on UKSI) speaking at a conference about the horrendously low number of taxon names that have never been mis-published or disputed in some way so can be guaranteed to be accurate - this was about scientific names, not just vernaculars.

We don't actually have a good solution to this issue. We could perhaps add a field for capturing the GBIF taxon ID, or the Catalogue of Life ID, to map to some global single taxonomic standard. But this would need to be done manually as UKSI doesn't contain these values and I doubt that all the names will just match up.

Is there a requirement to include EBMS data on the iRecord species pages and vice-versa?

**Comment by kitenetter** [comment]
@johnvanbreda do we have organism keys for all UKSI taxa now?

If so, either that or external_key should work for UKSI names I think.

I'm not aware of any requirement to include EBMS data, unless @DavidRoy knows otherwise.

**Comment by johnvanbreda** [comment]
> do we have organism keys for all UKSI taxa now?

We do (unless any manually added ones skipped the field), but it's not yet used in the Elasticsearch data.

**Comment by burkmarr** [comment]
Okay so for this issue - which only needs to map/chart UK data - it looks like I should use the `taxon.accepted_taxon_id` field to filter Indicia data for now. I will leave it for a day in case anyone else chips in here. But if that still looks like the consensus, II'll aim to apply a hotfix to update the species details page tomorrow.

**Comment by johnvanbreda** [comment]
I'd definitely use `taxon.accepted_taxon_id` at this point if you don't need to include the EBMS data.

**Comment by DavidRoy** [comment]
Agreed.  Only websites that share with iRecord for reporting are in scope for this, i.e. not EBMS currently

**Comment by kitenetter** [comment]
@burkmarr I see that the new maps are now available on the live site, which is excellent. We ought to make an announcement about this is there still any work left to do or is it ready to be announced?

Two loose ends:
1. Asian Hornet
The page for Asian Hornet shows no records, which I think is correct, as the records should be confidential. However, it does show a lot of photos of misidentified Asian Hornets. If the photos are from records that are currently identified as Asian Hornet then I think they should be excluded on grounds of confidentiality; but if the records have been redetermined or rejected than they should be excluded on those grounds.

2. Navigation to the new maps
There is currently no obvious way for people to find the new maps - you have to go to an individual species record page and then click on the button for the details page. I think we need to review this, and perhaps link to the new maps from the "SPecies maps" menu item, which currently goes [here](https://irecord.org.uk/species-maps), or else enable a way of going straight from a species name on the Explore pages to the details page.

I suggest we set up a new issue for this latter point - I will do that shortly.

**Comment by burkmarr** [comment]
@kitenetter - re https://github.com/BiologicalRecordsCentre/iRecord/issues/1286#issuecomment-1246589947 - I should fix the problem illustrated with the Peacock butterfly (see https://github.com/BiologicalRecordsCentre/iRecord/issues/1286#issuecomment-1243679077) before we announce it. I hope to do that today but am currently working on a potentially urgent problem with another project. I agree with setting up an new issue for the navigation. I don't know what will be involved in the image problem, but hopefully only an adjustment to the report which gets the images. If you want that to be fixed before we announce the updated page, then I suggest you leave that as part of this issue, but if it can wait, that can also go in a separte issue.

**Comment by burkmarr** [comment]
@kitenetter - re https://github.com/BiologicalRecordsCentre/iRecord/issues/1286#issuecomment-1246589947 - this could be related to the problem discussed here - https://github.com/BiologicalRecordsCentre/iRecord/issues/1286#issuecomment-1243679077. I'll check it out in more detail when I address that.

**Comment by kitenetter** [comment]
I won't announce anything just yet! We do need to clarify the handling of Asian Hornet before we announce anything, but that can wait until you have time to look into it.

**Comment by johnvanbreda** [comment]
For me, the Explore pages filtered for the 2 different taxon meaning IDs for Asian Hornet both work OK and both produce 411 records, so the filter is picking up both as I'd hope:
https://irecord.org.uk/all-records?filter-date_age=&filter-taxon_meaning_list=129053
https://irecord.org.uk/all-records?filter-date_age=&filter-taxon_meaning_list=134903 

The reason this works is both the NNSS and UKSI versions of Asian Hornet both have the same external key, which will get looked up using the taxon meaning ID and used for the filter. In fact there is a bit of extra functionality in the proxy code which converts a filter into Elasticsearch syntax. This makes it tolerate species in out of date lists which have the external key (i.e. the accepted name) now pointing to a synonym - see https://github.com/Indicia-Team/client_helpers/blob/master/ElasticsearchProxyHelper.php#L1249.

 The species summary pages now seem to be filtering on accepted name for the Elasticsearch requests which drive the maps - this doesn't work for the NNSS species as they are just lists of common names.

**Comment by burkmarr** [comment]
@kitenetter - I will switch the ES filter, as outlined above, to external key. That means that the same map will be returned whichever taxon_meaning is used - so long as they link to the same external key (as they do for Asian Hornet).

I've just looked at the Asian Hornet records in ES using Kibana and I'm a little confused about confidentiality. *All* the Asian Hornet records (regardless of their taxon_meaning_id) are marked as confidential. Currently I have ignored this flag when constructing the hectad map on this species details page. However the hectad map is not constructed from all the records. This is because many of the records (from both taxon_meaning_ids) appear to have their location blurred (to a 100 km square) and cannot therefore be used to build the hectad map.

**Comment by kitenetter** [comment]
@burkmarr my understanding is that all incoming Asian Hornet records are being flagged as confidential, but that there are some older ones that have not yet had a retrospective flag applied. I believe Steph is working with John to get them all flagged as confidential.

Confidential records should be excluded from any mapping.

**Comment by burkmarr** [comment]
@kitenetter - so the map for Asian Hornet will be blank. I guess I should add some text to the species details page that says that all confiential records are excluded from the map. I presume it's okay to use confidential records for the construction of the two temporal charts?

**Comment by kitenetter** [comment]
Asian Hornet is a special case due to its sensitivity, and to the fact that nearly all records are misidentifications. So for that one, and more generally, I'd be inclined to exclude all confidential records from any of the visualisations. AFAIK there aren't many confidential records in the system other than Asian Hornet.

**Comment by kitenetter** [comment]
I had a feeling that confidential records were being excluded from ES, sorry, I should have checked that first. In which case the only problem we have with the Asian Hornet records is that some of the older ones are not yet flagged as confidential (which I think Steph has already asked John about).

So there is probably no further action needed for the visualisations.

**Comment by burkmarr** [comment]
@johnvanbreda - thanks I hadn't realised that. The page configuration points to the es-occurrences endpoint.

**Comment by johnvanbreda** [comment]
That should automatically apply confidential filter for the reporting sharing mode then.

**Comment by kitenetter** [comment]
Just need to make the maps easier to find, see #1378 - closing this issue.
